File: DSA/LeetCode/Hashmaps/3160. Find the Number of Distinct Colors Among the Balls.py
from collections import defaultdict


class Solution:
    def queryResults(self, limit: int, queries: List[List[int]]) -> List[int]:
        
        
        index2Color = {}
        colorFreqMap = defaultdict(int)
        res = []

        for  [index, color] in queries:
            oldColor = False
            if  index in index2Color:
                oldColor = index2Color[index]
            oldColorFreq = colorFreqMap[oldColor]
            newColorFreq = colorFreqMap[color]
            if not oldColor:

                index2Color[index] = color
                colorFreqMap[color] =  newColorFreq + 1
            elif oldColor != color:


                if oldColorFreq == 1:
                    colorFreqMap.delete(oldColor)
                else:
                    colorFreqMap[oldColor] = oldColorFreq - 1
            

            index2Color[index] =  color
            colorFreqMap[color] = newColorFreq + 1

        

            res.append(colorFreqMap.size)
        return res
    

File: DSA/LeetCode/Hashmaps/3160. Find the Number of Distinct Colors Among the Balls.js
/**
 * @param {number} limit
 * @param {number[][]} queries
 * @return {number[]}
 */
var queryResults = function (limit, queries) {

    const index2Color = new Map();
    const colorFreqMap = new Map()
    let res = []

    for (q of queries) {
        const [index, color] = q

        let oldColor = index2Color.get(index)
        let oldColorFreq = colorFreqMap.get(oldColor) ?? 0
        let newColorFreq = colorFreqMap.get(color) ?? 0
        if (!oldColor) {

            index2Color.set(index, color)
            colorFreqMap.set(color, newColorFreq + 1)
        } else if (oldColor != color) {


            if (oldColorFreq == 1) {
                colorFreqMap.delete(oldColor)
            } else {
                colorFreqMap.set(oldColor, oldColorFreq - 1)
            }

            index2Color.set(index, color)
            colorFreqMap.set(color, newColorFreq + 1)

        }

        res.push(colorFreqMap.size)



    }

    return res

};